staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -280,36 +279,55 @@ func (qs *queueSet) StartRequest(ctx context.Context, hashValue uint64, flowDist
 	// ========================================================================
 	// Step 3:
 
-	// Set up a relay from the context's Done channel to the world
-	// of well-counted goroutines. We Are Told that every
-	// request's context's Done channel gets closed by the time
-	// the request is done being processed.
-	doneCh := ctx.Done()
-
 	// Retrieve the queueset configuration name while we have the lock
 	// and use it in the goroutine below.
 	configName := qs.qCfg.Name
 
-	if doneCh != nil {
+	// Set up a relay from the context's Done channel to the world
+	// of well-counted goroutines. We Are Told that every
+	// request's context's Done channel gets closed by the time
+	// the request is done being processed.
+	if ctx.Done() != nil {
 		qs.preCreateOrUnblockGoroutine()
 		go func() {
-			defer runtime.HandleCrash()
+			queueWaitCtx, cancel := getRequestWaitContext(ctx)
+			defer func() {
+				cancel()
+				runtime.HandleCrash()
+			}()
+
 			qs.goroutineDoneOrBlocked()
-			_ = <-doneCh
```

Comment:
@MikeSpreitzer ^

Reply:
The thing that unblocks this receive should call `qs.preCreateOrUnblockGoroutine()` before the send.  You are probably using a sender that does not do this.  You will have to change the test to use an appropriate sender.  The FakeEventClock can schedule work in the future; I think you will have to use that.

Reply:
Also, you rarely want to mix real clocks and fake clocks.  We use fake clocks in testing queueset, so that we can control time precisely and get repeatable tests and they run scenarios faster than real time.